db/db_filesnapshot.cc

```diff
@@ -154,6 +159,211 @@ Status DBImpl::GetCurrentWalFile(std::unique_ptr<LogFile>* current_log_file) {
+        flush_memtable = false;
+      }
+      live_wal_files.clear();
+    }
+  }
+
+  // this will return live_files prefixed with "/"
+  s = GetLiveFiles(live_files, &manifest_file_size, flush_memtable);
+
+  if (!GetIntProperty(DB::Properties::kMinLogNumberToKeep, &min_log_num)) {
+    return Status::InvalidArgument("cannot get the min log number to keep.");
+  }
+  // Between GetLiveFiles and getting min_log_num, flush might happen
+  // concurrently, so new WAL deletions might be tracked in MANIFEST. If we do
+  // not get the new MANIFEST size, the deleted WALs might not be reflected in
+  // the checkpoint's MANIFEST.
+  //
+  // If we get min_log_num before the above GetLiveFiles, then there might
+  // be too many unnecessary WALs to be included in the checkpoint.
+  //
+  // Ideally, min_log_num should be got together with manifest_file_size in
+  // GetLiveFiles atomically. But that needs changes to GetLiveFiles' signature
+  // which is a public API.
+  s = GetLiveFiles(live_files, &manifest_file_size, flush_memtable);
+  TEST_SYNC_POINT("CheckpointImpl::CreateCheckpoint:FlushDone");
+
+  TEST_SYNC_POINT("CheckpointImpl::CreateCheckpoint:SavedLiveFiles1");
+  TEST_SYNC_POINT("CheckpointImpl::CreateCheckpoint:SavedLiveFiles2");
+
+  if (s.ok()) {
+    s = FlushWAL(false /* sync */);
+  }
+
+  TEST_SYNC_POINT("CheckpointImpl::CreateCustomCheckpoint:AfterGetLive1");
+  TEST_SYNC_POINT("CheckpointImpl::CreateCustomCheckpoint:AfterGetLive2");
+
+  // if we have more than one column family, we need to also get WAL files
+  if (s.ok()) {
+    s = GetSortedWalFiles(live_wal_files);
+  }
+  if (!s.ok()) {
+    return s;
+  }
+
+  std::unique_ptr<FilesStorageInfoImpl> result{new FilesStorageInfoImpl};
+
+  size_t wal_size = live_wal_files.size();
+
+  // process live files, non-table, non-blob files first
+  std::string manifest_fname;
+  for (auto& live_file : live_files) {
+    if (!s.ok()) {
+      break;
+    }
+    uint64_t number;
+    FileType type;
+    bool ok = ParseFileName(live_file, &number, &type);
+    if (!ok) {
+      s = Status::Corruption("Can't parse file name. This is very bad");
+      break;
+    }
+    // we should only get sst, blob, options, manifest and current files here
+    assert(type == kTableFile || type == kBlobFile || type == kDescriptorFile ||
+           type == kCurrentFile || type == kOptionsFile);
+    assert(live_file.size() > 0 && live_file[0] == '/');
+
+    result->infos.emplace_back();
+    FileStorageInfo& info = result->infos.back();
+    info.relative_filename = live_file.substr(1);
+    info.directory = GetName();  // FIXME: support db_paths/cf_paths
+    info.file_number = number;
+    info.file_type = type;
+    if (type == kDescriptorFile) {
+      info.size = manifest_file_size;
+      info.trim_to_size = true;
+      assert(manifest_fname.empty());
+      manifest_fname = live_file;
+    } else if (type == kCurrentFile) {
+      info.size = 0;
```

Comment:
Didn't get this.

Reply:
See the comment on `FileStorageInfo::trim_to_size`

Reply:
Should we copy the content of the CURRENT file to a string instead?